src/core/package.go

```diff
@@ -134,29 +134,50 @@ func (pkg *Package) HasOutput(output string) bool {
 
 // RegisterOutput registers a new output file in the map.
 // Returns an error if the file has already been registered.
-func (pkg *Package) RegisterOutput(fileName string, target *BuildTarget) error {
+func (pkg *Package) RegisterOutput(state *BuildState, fileName string, target *BuildTarget) error {
 	pkg.mutex.Lock()
 	defer pkg.mutex.Unlock()
+
 	originalFileName := fileName
 	if target.IsBinary {
 		fileName = ":_bin_" + fileName // Add some arbitrary prefix so they don't clash.
 	}
+
 	if existing, present := pkg.Outputs[fileName]; present && existing != target {
-		if existing.IsFilegroup && !target.IsFilegroup {
-			// Update the existing one with this, the registered outputs should prefer non-filegroup targets.
-			pkg.Outputs[fileName] = target
-		} else if !target.IsFilegroup && !existing.IsFilegroup {
-			return fmt.Errorf("rules %s and %s in %s both attempt to output the same file: %s",
-				existing.Label, target.Label, pkg.Filename, originalFileName)
+		// Only local files are available as outputs to filegroups at this stage, so unless both targets are filegroups
+		// then the same output isn't allowed.
+		if !target.IsFilegroup || !existing.IsFilegroup {
+			// TODO(tiagovtristao): This condition exists for backwards compatibility and can be fully deleted when the FF is removed.
+			if state.Config.FeatureFlags.PackageOutputsStrictness || !target.IsFilegroup && !existing.IsFilegroup {
+				return fmt.Errorf("rules %s and %s in %s both attempt to output the same file: %s", existing.Label, target.Label, pkg.Filename, originalFileName)
+			}
+		}
+	}
+
+	if state.Config.FeatureFlags.PackageOutputsStrictness {
+		// Taking the new output file into account, check that if there are targets that output into a subdirectory, that subdirectory isn't created by another target.
+		outputs := make(map[string]*BuildTarget)
+		for f, t := range pkg.Outputs {
+			outputs[f] = t
+		}
+		outputs[fileName] = target
+		for filename, target := range outputs {
+			for dir := path.Dir(filename); dir != "."; dir = path.Dir(dir) {
+				if target2, present := outputs[dir]; present && target2 != target && !target.HasDependency(target2.Label.Parent()) {
+					return fmt.Errorf("Target %s outputs files into the directory %s, which is separately output by %s. You need to add a dependency to fix this", target.Label, dir, target2.Label)
+				}
+			}
```

Comment:
Not really comfortable with all this logic here, but I couldn't find a better way to abstract it out given that we should lock the pkg when we are trying to access its data, and it would conflict with the existing lock in this function